quill-engine/src/main/scala/io/getquill/ast/Ast.scala

```diff
@@ -221,113 +219,118 @@ case class Aggregation(operator: AggregationOperator, ast: Ast) extends Query {
+  override def bestQuat: Quat = a.bestQuat
 }
 
-case class DistinctOn(query: Ast, alias: Ident, body: Ast) extends Query {
-  def quat           = query.quat
-  def bestQuat: Quat = query.bestQuat
+final case class DistinctOn(query: Ast, alias: Ident, body: Ast) extends Query {
+  override def quat: Quat     = query.quat
+  override def bestQuat: Quat = query.bestQuat
 }
 
-case class Nested(a: Ast) extends Query {
-  def quat           = a.quat
-  def bestQuat: Quat = a.bestQuat
+final case class Nested(a: Ast) extends Query {
+  override def quat: Quat     = a.quat
+  override def bestQuat: Quat = a.bestQuat
 }
 
 //************************************************************
 
 final class Infix(val parts: List[String], val params: List[Ast], val pure: Boolean, val transparent: Boolean)(
   theQuat: => Quat
 ) extends Ast {
-  def quat: Quat     = theQuat
-  def bestQuat: Quat = quat
-  private def id     = Infix.Id(parts, params, pure, transparent)
+  override lazy val quat: Quat = theQuat
+  override def bestQuat: Quat  = quat
+  private val id               = Infix.Id(parts, params, pure, transparent)
 
-  def asTransparent = new Infix(this.parts, this.params, this.pure, true)(theQuat)
+  def asTransparent = new Infix(this.parts, this.params, this.pure, true)(quat)
 
-  override def equals(that: Any) =
+  override def equals(that: Any): Boolean =
     that match {
       case p: Infix => this.id == p.id
       case _        => false
     }
 
-  override def hashCode = id.hashCode()
+  override def hashCode: Int = id.hashCode()
   def copy(
     parts: List[String] = this.parts,
     params: List[Ast] = this.params,
     pure: Boolean = this.pure,
     transparent: Boolean = this.transparent,
     quat: Quat = this.quat
-  ) =
+  ): Infix =
     Infix(parts, params, pure, transparent, quat)
 }
 object Infix {
-  private case class Id(val parts: List[String], val params: List[Ast], val pure: Boolean, val transparent: Boolean)
+  private final case class Id(
+    parts: List[String],
+    params: List[Ast],
+    pure: Boolean,
+    transparent: Boolean
+  )
 
   def apply(parts: List[String], params: List[Ast], pure: Boolean, transparent: Boolean, quat: => Quat) =
     new Infix(parts, params, pure, transparent)(quat)
 
-  def unapply(i: Infix) = Some((i.parts, i.params, i.pure, i.transparent, i.quat))
+  def unapply(i: Infix): Option[(List[String], List[Ast], Boolean, Boolean, Quat)] =
+    Some((i.parts, i.params, i.pure, i.transparent, i.quat))
 }
 
-case class Function(params: List[Ident], body: Ast) extends Ast {
-  def quat           = body.quat
-  def bestQuat: Quat = body.bestQuat
+final case class Function(params: List[Ident], body: Ast) extends Ast {
+  override def quat: Quat     = body.quat
+  override def bestQuat: Quat = body.bestQuat
 }
 
 final class Ident private (val name: String)(theQuat: => Quat)(val visibility: Visibility) extends Terminal with Ast {
-  private lazy val computedQuat = theQuat
```

Comment:
We don't need this `computedQuat` variable. the overrided `def quat: Quat` method can be just a `lazy val`